### PR TITLE
[MDB IGNORE] Procedurally Generated Maints (Sorry no Gax in this PR just the tools)

### PR DESCRIPTION
**Pull request by Chubbygummibear**
After like 2 months of work I present the first iteration of my magnum opus, procedurally generated maint areas, and an entirely new map generation engine designed to fill designated areas with rooms (rectangular structures that you can decide what to do with). You can specify the behavior of these rooms from width and height, to what types of walls and floors they're constructed with and what sort of mobs and features it can spawn. Additionally you can even edit the items and mobs inside to rename them, change variables like factions, or alter their health/equipment/icons.

## Visual Demonstration of how the maint generator functions
The logic has been changed since this demo was made, but imagine that black rooms are space rooms, green are randomly themed, red and blue are prefabricated rooms. The order of operations is:

1. Get the dimensions of the overall area by aggregating valid procedural gen areas into one massive blob of tiles. From this it derives the dimensions of the area as min and max x and y values.
2. Send this data to my Rust function that first uses Binary Space Partition to cut up the area into subsections until a minimum subsection size is achieved for all subsections
3. Generate rooms in each subsection ranging in size from the minimum specified size, to the maximum subsection size
4. After BSP is done, it then runs the Random Room Placement algorithm to sprinkle more rooms of pre-made maint ruin sizes so that there's more ruins in the maints. It's worth noting that Random Room Placement will not respect BSP generated rooms and will just slap themselves over the top.

bsp rust demo

Here's a bad asesprite animation I made to show how it handles gax
bsp explaination

## This got testmerged and I want to play in the random maints but it's only on GAX :(((((
First of all, sorry you're not playing on Gax, the best map in the game. Secondly you can try to beg an admin to open the mapping controller subsystem and use the generate_backrooms proc to create The Backrooms z-level. 
**However be warned this will make the server shit itself and takes roughly 3 minutes to complete**

https://github.com/yogstation13/Yogstation/assets/46236974/d454582c-12cc-4fae-80ae-bf7bc6652862

## Can I use this to make an Ice Caves Dungeon Generator
Yeah it's just gonna take a bit of behavior changing and theme crafting

## Can I use this to create an entire procedurally generated station??????
Technically yes, but god have mercy on you for how much behavior specifying code that will require

## What does this area look like for mappers?
firefox_pEU8JQwzeN
image
you will need to create new areas of a shared subtype that use the same map_generator for them to cooperate and use one large continuous area during map generation

# Why is this good for the game?
Variety is the spice of life, and random maints will offer tons of variety in how antagonists interact with the round and offer something more dangerous for board assistants to explore. This was originally born of the idea that randomized maints would give bloodsuckers a better chance to really hide since the layout of the area can't possibly be committed to memory ahead of time.

# Testing
I've done a lot of personal testing, but I'm pretty sure this is going to require a lengthy testmerge period to iron out bugs, balance loot that appears, and expand on possible room themes to increase diversity

# Spriting
While I did add sprite code, I don't think I added any sprites that didn't already exist

# Wiki Documentation

This will probably require that I write a readme or page on how to use the generator because it's rather complicated and this isn't quite something that can be easily documented without trawling code. Also it might be best not to include exactly what can appear and loot tables for each room to discourage hunting for certain rooms because they can have a gamer item people want.

# Special thanks to my dev friends who helped me along this journey

- @ToasterBiome who, when i suggested the idea of proc maints said "when you code it". To which i said "bet" and then he proceeded to implement more and more requirements for it during development. You're the best worst dad I've ever had

- @LazennG who suggested many themes and feedback and is always a good springboard for ideas

- @Mqiib for suggesting themes and what specific loot each room should have. his suggestions resulted in the rooms themes being able to change their spawn pool dynamically before generating and alter the behavior of spawned items afterwords

- @cowbot92 for confirming that the rust_g.DLL was crashing the server and IT WASN'T MY FAULT

# Changelog

:cl:
rscadd: New map generator called a Dungeon Generator. Which is used for procedurally generating areas that require rooms or controllable square/rectangular areas. For when lavaland's blanket noise generator is too indirect
rscadd: Procedurally Generated maints, an always random area with highly varied loot and dangerous mobs inhabiting it. Don't enter unless you're absolutely sure you're equipped and know what you're doing
rscadd: The Backrooms
wip: yes
mapping: Multi-Z GaxStation with procedurally generated maints
mapping: New /area/ type for proc gen
experimental: YES
/:cl:

### CODE REVIEW

**Comment by LassiH** [comment]
we need this

**Comment by Yarinoi** [comment]
okay maybe gax is decent

**Comment by cowbot92** [comment]
image

**Comment by Chubbygummibear** [comment]
> okay maybe gax is decent

**Comment by patpol4** [comment]
no way

**Review by patpol4 — APPROVED** [review]

**Comment by neeshacark** [comment]
god bless autism

**Comment by MajManatee** [comment]
image

**Comment by Chubbygummibear** [comment]
adding a little demo of the maints and generate_backrooms proc

https://github.com/yogstation13/Yogstation/assets/46236974/d454582c-12cc-4fae-80ae-bf7bc6652862

**Comment by adamsong** [comment]
Doesn't monstermos crash on multiz?

**Comment by Chubbygummibear** [comment]
> Doesn't monstermos crash on multiz?

I think Biome said it only crashes as a result of trying to have atmos flow between z-levels freely through open floor tiles. however ours don't try to pass atmos, so i think that mitigates the crashing. it does also present a funny possibility that if a z-level below is getting plasma flooded and you're on an observation deck, you would be able to watch the chaos without any of it reaching you just 10 feet up

**Comment by Moja10** [comment]
Merge without testing

**Review by cuackles — APPROVED** [review]

**Comment by Chubbygummibear** [comment]
alright the rust code has been merged and conflicts have been resolved. @ToasterBiome it's ready to be tested i think

**Comment by PhantornRU** [comment]
Still worked?

File: code/controllers/subsystem/mapping.dm [hunk 1]
```diff
@@ -196,7 +196,6 @@ SUBSYSTEM_DEF(mapping)
 	var/list/space_ruins = levels_by_trait(ZTRAIT_SPACE_RUINS)
 	if (space_ruins.len)
 		seedRuins(space_ruins, CONFIG_GET(number/space_budget), list(/area/space), space_ruins_templates)
-	seedStation()
```
Comment (ToasterBiome) [inline code/controllers/subsystem/mapping.dm]: why is this removed?

File: code/modules/power/gravitygenerator.dm [hunk 2]
```diff
@@ -26,6 +26,7 @@ GLOBAL_LIST_EMPTY(gravity_generators) // We will keep track of this by adding ne
 	use_power = NO_POWER_USE
 	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF
 	var/sprite_number = 0
+	var/ztrait //Set to a valid ZTRAIT define to have the gravgen provide gravity to all of the zlevels with said trait. Ex: ZTRAIT_STATION
```
Comment (ToasterBiome) [inline code/modules/power/gravitygenerator.dm]: is this applicable anymore? we already have multiz gravity

File: code/modules/reagents/reagent_containers/pill.dm [hunk 3]
```diff
@@ -102,8 +102,10 @@
 	name = "adminordrazine pill"
 	desc = "It's magic. We don't have to explain it."
 	icon_state = "pill16"
-	volume = 50
-	list_reagents = list(/datum/reagent/medicine/adminordrazine = 50)
+	volume = 5
```
Comment (ToasterBiome) [inline code/modules/reagents/reagent_containers/pill.dm:105]: why is this changed?

File: config/maps.txt [hunk 4]
```diff
@@ -44,7 +44,7 @@ endmap
 # Debug-only maps.
 
 map multiz_debug
-	disabled 
+
```
Comment (ToasterBiome) [inline config/maps.txt]: redisable

File: yogstation.dme [hunk 5]
```diff
@@ -788,9 +789,18 @@
 #include "code\datums\looping_sounds\weather.dm"
 #include "code\datums\mapgen\_MapGenerator.dm"
 #include "code\datums\mapgen\CaveGenerator.dm"
+#include "code\datums\mapgen\DungeonGenerator.dm"
 #include "code\datums\mapgen\biomes\_biome.dm"
 #include "code\datums\mapgen\Cavegens\IcemoonCaves.dm"
 #include "code\datums\mapgen\Cavegens\LavalandGenerator.dm"
+#include "code\datums\mapgen\Dungeon_Generators\Dungeon_Room.dm"
```
Comment (ToasterBiome) [inline yogstation.dme]: you are ass for making these folders capitalized

**Review by ToasterBiome — CHANGES_REQUESTED** [review]

File: code/modules/reagents/reagent_containers/pill.dm (hunk 3 above)
Comment (Chubbygummibear) [reply code/modules/reagents/reagent_containers/pill.dm:105]: because it sometimes appears in the medbay rooms and i didn't want people getting 50u of adminordrazine when as far as i can tell, 1u will do the full rejuvenate

File: code/modules/power/gravitygenerator.dm (hunk 2 above)
Comment (Chubbygummibear) [reply code/modules/power/gravitygenerator.dm]: probably not

File: yogstation.dme (hunk 5 above)
Comment (Chubbygummibear) [reply yogstation.dme]: uncapitalized

File: config/maps.txt (hunk 4 above)
Comment (Chubbygummibear) [reply config/maps.txt]: ok

File: code/controllers/subsystem/mapping.dm (hunk 1 above)
Comment (Chubbygummibear) [reply code/controllers/subsystem/mapping.dm]: why did you remove yourself from my life, dad?

**Review by ToasterBiome — APPROVED** [review]
